### ISSUE TITLE
Trying to register an open generic class with a constructor with an open generic interface as a parameter.

### ISSUE THREAD

**Issue by dworthem440** [simpleinjector/SimpleInjector#768]
In ASP.NET Web Forms project I have an open generic class with a constructor with an open generic interface as parameter.

``` vb
Public Class LogRouterApi(Of T As Class)
    Private ReadOnly _logger As IEventLogger(Of T)
    Public Sub New(logger As IEventLogger(Of T))
        _logger = logger
    End Sub
End Class
```

I have page classes like the following:

``` vb
Partial Class shhhhh
    Inherits System.Web.UI.Page
    <Import>
    Public Property _logger As LogRouterApi(Of shhhhh)
End Class    
```

I tried to register these with:

``` vb
container.Register(GetType(IEventLogger(Of)), GetType(EventLogger(Of)).Assemby)
container.Register(GetType(LogRouterApi(Of)), GetType(LogRouterApi(Of)).Assembly)
```

The problem is that `IEventViewer(Of T)` is implemented by `EventView(Of T)` but has 2 public constructors so that registration does not work. It works for `LogRouterApi(Of)`

I tried something like:

``` vb
container.Register(Of IEventLogger(Of ))(
    Function()
        Return New EventLogger(Of)()
    End Function, Lifestyle.Scoped)
```

But this does not compile.

The following works but it is not with open-generic type parameter. This was with the open generic registration of LogRouterApi(Of)

``` vb
container.Register(Of IEventLogger(Of shhhhh))(
    Function()
        Return New EventLogger(Of shhhhh)()
    End Function, Lifestyle.Scoped)
```

Any suggestions would be appreciated so I can register both as open generic types.

**Comment by dotnetjunkie** [comment]
I'm trying to understand your question, and I'm finding it hard to follow your question. Perhaps because I'm reading this in the browser, but would you mind reformatting your question? It also unclear to ne what the relationship is between your problem and Web Forms. Would you mind producing a minimal reproduce example that shows your problem with the least amount of code?

**Comment by dotnetjunkie** [comment]
Hi @dworthem440 ,

The `Container.Register(Type, Assembly[])` overload "Registers all concrete, **non-generic,** public and internal types in the given set of assemblies that implement the given openGenericServiceType". As you wish to map `IEventLogger<T>` to `EventLogger<T>`, you need the following overload:

``` vb
container.Register(GetType(IEventLogger(Of)), GetType(EventLogger(Of)))
```

Now about `EventLogger<T>` having two constructors, there are a few path you can walk here:

* Remove one of `EventLogger<T>`'s constructors. Having [multiple constructors is an anti-pattern](https://blogs.cuttingedge.it/steven/posts/2013/di-anti-pattern-multiple-constructors/). But if you didn't define that type, you can do the following two things:
* You can override Simple Injector's [constructor resolution behavior](https://simpleinjector.readthedocs.io/en/latest/extensibility.html#overriding-constructor-resolution-behavior).
* or you can create a new type (with one constructor) that derives from `EventLogger<T>` and register that.
